### ISSUE TITLE
rows occasionally arrive out-of-order unless I set parallel:1 

### ISSUE THREAD

**Issue by DrPaulBrewer** [adaltas/node-stream-transform#26]
I am finding that output rows can arrive at the filesystem in a different order than input rows if parallel is left at the default of 100.  My transformation function is not asynchronous, though, its just some deterministic math.  

With a .csv input file length ~ 10,000 rows I am seeing about 15 rows that are in a different order in the output than the input. This was discovered by running a transformation twice and using `diff --context file1.csv file2.csv` on the output files.

I set `{parallel: 1}` and the rows arrive in order again.  Repeating ~10 times doesn't break the order.  I explicitly set `{parallel:100}` and the issue immediately presents itself anew.

I think this is an unavoidable issue with parallelization, not something to fix.

It might suggest a bit more documentation of how/when to turn parallel off.

I can provide a reproducible example a little later if that is useful.

**Comment by wdavidw** [comment]
Or maybe it shall be sequential (`parallel: 1`) by default.

**Comment by DrPaulBrewer** [comment]
Whether or not parallel is useful possibly depends on whether the transformation has more or less throughput than the filesystem.  

I thought it would only be on for async transformation function -- which should be easy enough to check if the return is a Promise --  but did not check your source.  So for synchronous math operations on the columns, ideally this would not be happening.

**Comment by wdavidw** [comment]
> Whether or not parallel is useful possibly depends on whether the transformation has more or less throughput than the filesystem.
No, this is Node.js, it is single threaded. It depends wether or not you have async io and, if so, how many of them you want in parallel

**Comment by DrPaulBrewer** [comment]
The single thread isn't the issue.  A common async io application is a REST request.  A REST request could be anything from all the tweets on a topic in the last second to accessing website analytics to remote in-memory caching.  But the available throughput on an external service is finite.  Even something like Google Drive, you'll start getting 429s if  requests are too aggressive,  Typical Google Drive API quota is 1,000 queries per user per 100 seconds so more than 5-10 requests at a time may be too many.  

Because of diversity of applications, and need for external facts and limits that can vary, I don't see any automated way of choosing the parallelization that would work well for everyone.

**Comment by wdavidw** [comment]
We think to agree on the sync/async nature of transformations but I don't get the point of this issue. If synchronous, then rows are kept in order, if asynchronous, then ordering can't be respected if multiple transformations are scheduled in parallel.

**Comment by DrPaulBrewer** [comment]
Yes, conversation has drifted.

Although as you say it should be impossible, I have synchronous math
example that produces out of order rows with stream-transform unless the
option {parallel:1} is explicitly set in transformer options.

As for an example the specific code is the gini-csv.js calculator in GitHub
drpaulbrewer/gini-csv; I can supply an input file and a command later when
I have more time.  The current version sets parallel to 1 but this can be
changed near the top of the file to blank object to induce the out of order
rows.

So the issue is that synchronous transformation function does not in fact
behave as described.

**Comment by wdavidw** [comment]
Ok, but my point was that if the code is 100% synchronous, then there is no optimization to expect in set parallel to a value above 1. This is why to me, synchronous is a synonym of sequential (parallel = 1). But you are right, we shall expect the order to be maintained if `parallel is above 1. Also, I still think that having `parallel` default to `1` make more sense than an arbitrary value of `100`. I am just afraid to break people code, despite using a major version release.

**Comment by wdavidw** [comment]
I have reproduce the situation inside a test:

```coffee
    it 'respect order', ->
      generate length: 1000, objectMode: true, seed: 1, columns: 2, (err, source) ->
        transform source, ((record) -> record), (err, result) ->
          source.should.eql result
```
 I will look how hard it is to implement it. I haven't been into this code for a long time.

**Comment by wdavidw** [comment]
Could you please show me how you call `generate`, it seems like you are using it by passing the dataset as an argument (in opposition to a stream). If this is a case, the fix is easy and I have it.

**Comment by wdavidw** [comment]
Version 2.0.1 shall fix the issue if you confirm the above message.

**Comment by DrPaulBrewer** [comment]
I have external data and so the .csv file in put already exists on disk and
I do not need to call generate.

I convert a data file to a stream with nodejs createReadStream.  The entire
program becomes a lot of setup then a chain of .pipe statements.  This
usage is only briefly mentioned in your docs but is the typical way to use
streams in general in nodejs.

My program worked without any additional coding on your part.  How you want
to fix or address the order issue is up to you.

 I think making parallel equal 1 is a good idea, and then briefly also
mention the parallel/order trade-off in docs.

Most popular uses of stream I have seen in programming preserve order of
records (e g. Pipes in bash, tcp/IP sockets, nodejs createReadStrwam) and
so could be a bit of surprise to new users.

**Comment by DrPaulBrewer** [comment]
If you want your generate and transform to have similar timing properties to mine, I think it would reduce to these specs:

~10,000 rows of  ~40 numeric columns, from Math.random()

Then in your transform, calculate something O(COL^2) like the sum of data[i]*data[j] for i!=j 

This is a similar amount of work for the cpu/disk to what I am doing,  a lot of Sum[Abs[data[i]-data[j]]] 

In my case, this would yield 10-15 row misplacements. I detect by running twice and using `diff`

**Comment by DrPaulBrewer** [comment]
I noticed your commit, changing `process.nextTick()` to `setImmediate()`.  [Stack overflow seems to like setImmediate better](https://stackoverflow.com/questions/15349733/setimmediate-vs-nexttick)

I'll have some time test in about 24 hours.

**Comment by wdavidw** [comment]
Could we close the issue as it shall be fixed with version 2.0.1 ?

**Comment by DrPaulBrewer** [comment]
Yes
